src/Constraints.hs

```diff
@@ -161,26 +161,22 @@ checkForConflict :: TypeMappings -> Constraint -> String -> Ty -> Either Unifica
 checkForConflict mappings constraint name otherTy =
   let (Constraint _ _ xobj1 xobj2 ctx  _) = constraint
       found = recursiveLookup mappings name
-  in
-    if doesTypeContainTyVarWithName name otherTy
-    then Left (UnificationFailure constraint mappings)
-    else
-      case found of --trace ("CHECK CONFLICT " ++ show constraint ++ " with name " ++ name ++ ", otherTy: " ++ show otherTy ++ ", found: " ++ show found) found of
+  in case found of --trace ("CHECK CONFLICT " ++ show constraint ++ " with name " ++ name ++ ", replaced: " ++ show replaced ++ ", found: " ++ show found) found of
```

Comment:
Correct me if I'm wrong, but I feel like the removed error check here actually did catch an error, namely the unification of something like `(T a) <=> a`. I actually think the problem is that `concretizeType` in Concretize.hs, in the Sumtype case might be the true culprit. That's where it sends both the looked up types to unification (via `instantiateGenericSumtype`) without doing anything to the type variables. They must be renamed using `renameVarTys` or similarly, I believe. What do you think?

Reply:
Thanks for calling this out! I did some investigating, and the check was introduced in commit b1248ac37 back in 2018. Looks like it was introduced to solve an issue with local variable names:

```diff
-*** Defining a local variable messes up unqualified lookup of function with the same name, e.g. (let [words (words &s)] ...). Infinite loop now!
+*** Defining a local variable messes up unqualified lookup of function with the same name, e.g. (let [words (words &s)] ...)
```

After this PR the behavior is the same as after b1248ac37, namely the above still doesn't work, but it won't trigger an infinite unification loop. 

We still perform the check, the difference is we just use it to rename variables with conflicting names instead of returning a unification error, which still prevents the infinite loop (see line 214):

```haskell
replaced = if doesTypeContainTyVarWithName name otherTy
                   then replaceConflicted name otherTy
                   else otherTy
```

I'm fairly sure this ensures behavioral parity, but I could be wrong about that--do you have an example case that might break?

I do think that if there's a better way to solve it elsewhere we should take that route--it'd involve more than sumtypes though, as this particular issue also arises in the case of interfaces. When we return a unification error for that check we can't unify anything that coincidentally shares a variable name, making it impossible to unify something like `interface a -> a` with `implementation (Ref a) -> (Ref a)` which should be unifiable.

Reply:
All that said, it's possible the better approach is to call some function like `renameTyVars` in `Primitives` where we constrain the interface and implementation rather than replacing them here.

Actually, looking into it, it looks like the correct approach here is what you're suggesting, Erik. 

To solve the interface unification problem, we should do something like the following in the interfaces code instead:

```
unifiedInterface = replaceTyVars interface . (unifySignatures interface implementation)
solve interface unifiedInterface
```

Then, as you noted Erik, the original check should stand, and the real issue lies somewhere in sumtypes.

Reply:
I'm going to give the above a try and unsolve #521 which is likely separate issue.

Reply:
> I'm going to give the above a try and unsolve #521 which is likely separate issue.

Ah nope--that still doesn't work since the variables may still be the same. We'd have to work around the check regardless...

Reply:
@eriksvedang I can't find any cases that removing the Type var check effects, but I've added ff0596e which takes a more conservative approach--instead of removing that check, we just bypass it for the special cases of interface registration. This should ensure we don't break anything by accident.

This makes the change more focused, and we'll have to solve #521 elsewhere.